File: packages/dukan_core/lib/domain/insights.dart
/// Insight rules. Mirrors server/dukan/domain/insights.py — the SAME rules run
/// in both languages. Pure: no imports. Quantities are integer minor units
/// (see Money invariant); an insight is a decision, its wording lives in the UI.

enum InsightSeverity { info, warning, critical }

/// Suggested reorder quantity (minor units) to bring stock back to twice the
/// reorder point, or 0 when stock still covers lead + safety demand.
///
/// reorderPoint = avgDailySales × (leadTimeDays + safetyDays).
int reorderSuggestion({
  required int onHandMinor,
  required int avgDailySalesMinor,
  int leadTimeDays = 7,
  int safetyDays = 3,
}) {
  if (avgDailySalesMinor <= 0) return 0;
  final reorderPoint = avgDailySalesMinor * (leadTimeDays + safetyDays);
  if (onHandMinor > reorderPoint) return 0;
  final qty = reorderPoint * 2 - onHandMinor;
  return qty < 0 ? 0 : qty;
}

/// Stock that is on hand but has not sold within [deadAfterDays] — capital tied
/// up in slow inventory.
bool isDeadStock({
  required int onHandMinor,
  required int daysSinceLastSale,
  int deadAfterDays = 30,
}) =>
    onHandMinor > 0 && daysSinceLastSale >= deadAfterDays;

/// How urgent a customer's outstanding balance is, relative to their credit
/// limit: at/over the limit is critical, ≥80% is a warning.
InsightSeverity debtSeverity({
  required int balanceMinor,
  required int creditLimitMinor,
}) {
  if (balanceMinor <= 0 || creditLimitMinor <= 0) return InsightSeverity.info;
  if (balanceMinor >= creditLimitMinor) return InsightSeverity.critical;
  if (balanceMinor * 10 >= creditLimitMinor * 8) return InsightSeverity.warning;
  return InsightSeverity.info;
}

File: server/dukan/domain/insights.py
"""Insight rules. Mirrors packages/dukan_core/lib/domain/insights.dart — the
SAME rules in both languages. Pure: no imports beyond the enum. Quantities are
integer minor units; an insight is a decision, its wording lives in the UI.
"""

from __future__ import annotations

from enum import StrEnum


class InsightSeverity(StrEnum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


def reorder_suggestion(
    *,
    on_hand_minor: int,
    avg_daily_sales_minor: int,
    lead_time_days: int = 7,
    safety_days: int = 3,
) -> int:
    """Suggested reorder quantity (minor units) to reach twice the reorder
    point, or 0 when stock still covers lead + safety demand."""
    if avg_daily_sales_minor <= 0:
        return 0
    reorder_point = avg_daily_sales_minor * (lead_time_days + safety_days)
    if on_hand_minor > reorder_point:
        return 0
    qty = reorder_point * 2 - on_hand_minor
    return qty if qty > 0 else 0


def is_dead_stock(
    *, on_hand_minor: int, days_since_last_sale: int, dead_after_days: int = 30
) -> bool:
    """On hand but unsold within `dead_after_days` — capital tied up in slow stock."""
    return on_hand_minor > 0 and days_since_last_sale >= dead_after_days


def debt_severity(*, balance_minor: int, credit_limit_minor: int) -> InsightSeverity:
    """Urgency of a customer's balance vs their credit limit: at/over = critical,
    >= 80% = warning."""
    if balance_minor <= 0 or credit_limit_minor <= 0:
        return InsightSeverity.INFO
    if balance_minor >= credit_limit_minor:
        return InsightSeverity.CRITICAL
    if balance_minor * 10 >= credit_limit_minor * 8:
        return InsightSeverity.WARNING
    return InsightSeverity.INFO
